library/src/main/scala/scala/scalajs/runtime/RuntimeLong.scala

```diff
@@ -182,38 +182,93 @@ final class RuntimeLong(val lo: Int, val hi: Int)
-    else
-      new RuntimeLong(0, lo << n)
+  def <<(n: Int): RuntimeLong = {
+    /* This should *reasonably* be:
+     *   val n1 = n & 63
+     *   if (n1 < 32)
+     *     new RuntimeLong(lo << n1, if (n1 == 0) hi else (lo >>> 32-n1) | (hi << n1))
+     *   else
+     *     new RuntimeLong(0, lo << n1)
+     *
+     * Replacing n1 by its definition, we have:
+     *   if (n & 63 < 32)
+     *     new RuntimeLong(lo << (n & 63),
+     *         if ((n & 63) == 0) hi else (lo >>> 32-(n & 63)) | (hi << (n & 63)))
+     *   else
+     *     new RuntimeLong(0, lo << (n & 63))
+     *
+     * Since the values on the rhs of shifts are always in arithmetic mod 32,
+     * we can get:
+     *   if (n & 63 < 32)
+     *     new RuntimeLong(lo << n, if ((n & 63) == 0) hi else (lo >>> -n) | (hi << n))
+     *   else
+     *     new RuntimeLong(0, lo << n)
+     *
+     * The condition `n & 63 < 32` is equivalent to
+     *   (n & 63) & 32 == 0
+     *   n & (63 & 32) == 0
+     *   n & 32 == 0
+     *
+     * In the then part, we have `n & 32 == 0` hence `n & 63 == n & 31`:
+     *   new RuntimeLong(lo << n, if ((n & 31) == 0) hi else (lo >>> -n) | (hi << n))
+     *
+     * Consider the following portion:
+     *   if ((n & 31) == 0) hi else (lo >>> -n) | (hi << n)
+     * When (n & 31) == 0, `hi == (hi << n)` and therefore we have
+     *   (if ((n & 31) == 0) 0 else (lo >>> -n)) | (hi << n)
+     *
+     * The left part of the |
+     *   if ((n & 31) == 0) 0 else (lo >>> -n)
+     * has the following branchless version:
+     *   lo >>> 1 >>> (31-n)
+     * Indeed, when `n & 31 == 0, we have
+     *   lo >>> 1 >>> 31 == 0
+     * and when `n & 31 != 0`, we know that ((31-n) & 31) < 31, and hence we have
+     *   lo >>> 1 >>> (31-n) == lo >>> (1+31-n) == lo >>> (32-n) == lo >>> -n
+     *
+     * Was it good? We have traded
+     *   if ((n & 31) == 0) hi else (lo >>> -n) | (hi << n)
+     * for
+     *   (lo >>> 1 >>> (31-n)) | (hi << n)
+     * When (n & 31) != 0, which is the common case, we have traded a test
+     * `if ((n & 31) == 0)` for one additional constant shift `>>> 1`. That's
+     * probably worth it performance-wise. The code is also shorter.
+     *
+     * Summarizing, so far we have
+     *   if (n & 32 == 0)
+     *     new RuntimeLong(lo << n, (lo >>> 1 >>> (31-n)) | (hi << n))
+     *   else
+     *     new RuntimeLong(0, lo << n)
+     *
+     * If we distribute the condition in the lo and hi arguments of the
+     * constructors, we get a version with only one RuntimeLong output, which
+     * avoids reification as records by the optimizer, yielding shorter code.
+     * It is potentially slightly less efficient, except when `n` is constant,
+     * which is often the case anyway.
+     *
+     * Finally we have:
+     */
+    new RuntimeLong(
+        if ((n & 32) == 0) lo << n else 0,
+        if ((n & 32) == 0) (lo >>> 1 >>> (31-n)) | (hi << n) else lo << n)
```

Comment:
It's actually a good question. I've spent hours trying to decide how far to go. For example, there's a branchless version of the first `if` which is `(lo << n) & ~((n << 26) >> 31)`. I have stopped when reasoning was not enough to justify performance gains, and when changes would increase code size.

In this case, performance-wise, I have no idea what is better. I would have to look at the generated assembly (e.g., the JIT might already generate branchless code for the lo part). Code size-wise, your snippet will be longer.